File: src/main/java/coyote/wdm/online/S3NamespaceContext.java
package coyote.wdm.online;

import javax.xml.namespace.NamespaceContext;
import java.util.Iterator;

import static java.util.Collections.emptyIterator;
import static java.util.Collections.singletonList;

public class S3NamespaceContext implements NamespaceContext {

  private static final String S3_BUCKET_LIST_NS = "http://doc.s3.amazonaws.com/2006-03-01";
  private static final String S3_PREFIX = "s3";

  @Override
  public String getNamespaceURI(String prefix) {
    if (S3_PREFIX.equals(prefix)) {
      return S3_BUCKET_LIST_NS;
    }
    throw new IllegalArgumentException("Unsupported prefix");
  }

  @Override
  public String getPrefix(String namespaceURI) {
    if (S3_BUCKET_LIST_NS.equals(namespaceURI)) {
      return S3_PREFIX;
    }
    throw new IllegalArgumentException("Unsupported namespace URI");
  }

  @Override
  public Iterator<String> getPrefixes(String namespaceURI) {
    if (S3_BUCKET_LIST_NS.equals(namespaceURI)) {
      return singletonList(S3_PREFIX).iterator();
    } else {
      return emptyIterator();
    }
  }

}

File: src/main/java/coyote/wdm/online/S3NamespaceContext.kt
package coyote.wdm.online

import java.util.*
import javax.xml.namespace.NamespaceContext

/**
 * Namespace context for S3 buckets.
 */
class S3NamespaceContext : NamespaceContext {
    override fun getNamespaceURI(prefix: String): String {
        if (S3_PREFIX == prefix) {
            return S3_BUCKET_LIST_NS
        }
        throw IllegalArgumentException("Unsupported prefix")
    }

    override fun getPrefix(namespaceURI: String): String {
        if (S3_BUCKET_LIST_NS == namespaceURI) {
            return S3_PREFIX
        }
        throw IllegalArgumentException("Unsupported namespace URI")
    }

    override fun getPrefixes(namespaceURI: String): Iterator<String?> {
        return if (S3_BUCKET_LIST_NS == namespaceURI) {
            listOf(S3_PREFIX).iterator()
        } else {
            Collections.emptyIterator<String>()
        }
    }

    companion object {
        private const val S3_BUCKET_LIST_NS = "http://doc.s3.amazonaws.com/2006-03-01"
        private const val S3_PREFIX = "s3"
    }
}